src/jit/lsraxarch.cpp

```diff
@@ -2502,27 +2502,62 @@ void LinearScan::TreeNodeInfoInitHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree,
     {
         SetContainsAVXFlags(true, 32);
     }
-    info->srcCount += GetOperandInfo(intrinsicTree->gtOp.gtOp1);
-    if (intrinsicTree->gtGetOp2IfPresent() != nullptr)
+    GenTree* op1   = intrinsicTree->gtOp.gtOp1;
+    GenTree* op2   = intrinsicTree->gtOp.gtOp2;
+    info->srcCount = 0;
+
+    if (op1 != nullptr)
+    {
+        if (op1->OperIsList())
+        {
+            for (GenTreeArgList* list = op1->AsArgList(); list != nullptr; list = list->Rest())
+            {
+                info->srcCount += GetOperandInfo(list->Current());
+            }
+        }
+        else
+        {
+            info->srcCount += GetOperandInfo(op1);
+        }
+    }
+
+    if (op2 != nullptr)
     {
-        info->srcCount += GetOperandInfo(intrinsicTree->gtOp.gtOp2);
+        info->srcCount += GetOperandInfo(op2);
     }
 
-#ifdef _TARGET_X86_
-    if (intrinsicTree->gtHWIntrinsicId == NI_SSE42_Crc32)
+    switch (intrinsicID)
     {
-        // CRC32 may operate over "byte" but on x86 only RBM_BYTE_REGS can be used as byte registers.
-        //
-        // TODO - currently we use the BaseType to bring the type of the second argument
-        // to the code generator. May encode the overload info in other way.
-        var_types srcType = intrinsicTree->gtSIMDBaseType;
-        if (varTypeIsByte(srcType))
+        case NI_SSE41_BlendVariable:
         {
-            LocationInfoListNode* op2Info = useList.GetSecond(INDEBUG(intrinsicTree->gtGetOp2()));
-            op2Info->info.setSrcCandidates(this, RBM_BYTE_REGS);
+            if (!compiler->canUseVexEncoding())
+            {
+                // SSE4.1 blendv* hardcode the mask vector in XMM0
+                info->internalFloatCount = 1;
+                info->setInternalCandidates(this, RBM_XMM0);
```

Comment:
> Does "Right" means that you used setDstCandidates? Anyway, it's supposed to be setSrcCandidates, not sure why I wrote Dst instead.

@mikedn Here I tried `setSrcCandidates` and `setDstCandidates`, but op3 cannot get XMM0 neither.
```cpp
LocationInfoListNode* op3Info = useList.Begin()->Next()->Next();
op3Info->info.setSrcCandidates(this, RBM_XMM0);
```

Reply:
Hmm, I got your branch to take a closer look. It looks to me that the above code does actually work fine. The problem is that sometimes you may end up with xmm0 being also assigned as the destination register of `blendvps`. Since `dst` and `src1` are the same the usual solution is to first copy the value from `src1` to `dst`:
```asm
movaps xmm0, xmmA
blendvps xmm0, xmmB, xmm0
```
This is obviously incorrect because the mask that's supposed to be in `xmm0` is destroyed by the copy.

So there's no way for the `blendvps` destination to be `xmm0`. You could exclude `xmm0` by adding `info->setDstCandidates(this, allSIMDRegs() & ~RBM_XMM0);`.

I think it should also be possible to set `isDelayFree` on `op3Info->info` but that doesn't seem to work. I haven't looked too closely at this issue but I think that the fact that this node has 3 operands is a bit of a problem. This needs additional investigation.

Reply:
`isDelayFree` works fine, you just need to also add `info->hasDelayFreeSrc = true;`.

Reply:
Thank you so much for the help!